[Multiple articles] Restructure table of contents (#150)
microsoft/vcpkg-docs: vcpkg/users/buildsystems/cmake-integration.md

---
title: vcpkg in CMake projects

> description: Integrate vcpkg into a CMake project using Visual Studio, Visual Studio Code, a terminal, or other IDEs.

Reviewer: ```suggestion
description: Integrate vcpkg into a CMake project using a terminal, Visual Studio, Visual Studio Code, or other IDEs.
```

author: vicroms
ms.author: viromer
ms.prod: vcpkg

> ms.date: 11/30/2022

Reviewer: Probably worth updating the date at this point.

ms.topic: conceptual
---
# vcpkg in CMake projects

> vcpkg offers seamless integration with CMake to make installed packages available automatically in
> your CMake projects. The mecahnism in which vcpkg integrates is by providing a CMake toolchain file.

Reviewer 2: ```suggestion
vcpkg offers seamless integration with CMake to make installed packages available in
your projects automatically. The mechanism in which vcpkg integrates is by providing a CMake toolchain file.
```

The first time CMake configures a project, it runs internal search routines to locate a viable
[toolchain](<https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html#id9>) (compiler,
linker, etc.). This search happens within the
`[project()](https://cmake.org/cmake/help/latest/command/project.html#command:project)` function in
your `CMakeLists.txt`.

To customize the toolchain selection process, CMake supports using custom CMake-language scripts,
known as toolchain files. A toolchain file is specified by setting the `CMAKE_TOOLCHAIN_FILE`

> variable. CMake evaluates the contents of the provided toolchain script, and sets variable definitions, paths to
> required build tools, and other build parameters, such as cross-compilation flags accordingly.

Reviewer 2: ```suggestion
variable. CMake evaluates the contents of the provided toolchain script and sets variable definitions, paths to
required build tools, and other build parameters, such as cross-compilation flags, accordingly.
```

> When users set `CMAKE_TOOLCHAIN_FILE` to use the vcpkg toolchain
> (`<vcpkg-root>/scripts/buildsystems/vcpkg.cmake`). vcpkg takes advantage of the toolchain file
> mechanism to inject code to integrate with built-in CMake functions transparently to the user.

Reviewer: ```suggestion
When you set `CMAKE_TOOLCHAIN_FILE` to use the vcpkg toolchain
(`<vcpkg-root>/scripts/buildsystems/vcpkg.cmake`), vcpkg takes advantage of the toolchain file
mechanism to inject code to integrate with built-in CMake functions transparently to you.
```

> Users can still use a toolchain file to configure thei own toolsets by using the

Reviewer 2: ```suggestion
Users can still use a toolchain file to configure their own toolsets by using the
```

> Users can still use a toolchain file to configure their own toolsets by using the
> []`VCPKG_CHAINLOAD_TOOLCHAIN_FILE`](../users/triplets.md#VCPKG_CHAINLOAD_TOOLCHAIN_FILE) triplet variable.

Reviewer: ```suggestion
You can still use a toolchain file to configure your own toolsets by using the
[]`VCPKG_CHAINLOAD_TOOLCHAIN_FILE`](../users/triplets.md#VCPKG_CHAINLOAD_TOOLCHAIN_FILE) triplet variable.
```

Th vcpkg integration works differently depending on the operation mode you're using:

In [classic mode](../classic-mode.md), vcpkg sets CMake search paths appropriately to make 
installed packages available via the `find_package()`, `find_library()`, and `find_path()` functions.

In [manifest mode](../manifests.md), in addition to the above, the toolchain detects manifest files
(`vcpkg.json` files) and runs `vcpkg install` to automatically acquire the project's dependencies.

> Because the toolchain file is evaluated during the `project()` call. All CMake-level variables that
> modify a vcpkg setting must be set before the first call to `project()`. It may also be necessary to
> reconfigure your CMake project if you modify any vcpkg setting that results in any [ABI
> hashes](../binarycaching.md#abi-hash) changing.

Reviewer: ```suggestion
Because the toolchain file is evaluated during the `project()` call, all CMake-level variables that
modify a vcpkg setting must be set before the first call to `project()`. It may also be necessary to
reconfigure your CMake project if you modify any vcpkg setting that results in [ABI
hash](../binarycaching.md#abi-hash) changes.
```

> In [manifest mode](../manifests.md), in addition to the above, the toolchain detects manifest files
> (`vcpkg.json` files) and runs `vcpkg install` to acquire the dependencies of the project automatically.
> 
> Because the toolchain file is evaluated during the `project()` call. All CMake-level variables that
> modify a vcpkg setting must be set before the first call to `project()`. It may also be necessary to
> reconfigure your CMake project if you modify any vcpkg setting that results in any [ABI
> hashes](../binarycaching.md#abi-hash) to change.

Reviewer 2: ```suggestion
In [manifest mode](../manifests.md), in addition to the above, the toolchain detects manifest files
(`vcpkg.json` files) and runs `vcpkg install` to automatically acquire the project's dependencies.

Because the toolchain file is evaluated during the `project()` call. All CMake-level variables that
modify a vcpkg setting must be set before the first call to `project()`. It may also be necessary to
reconfigure your CMake project if you modify any vcpkg setting that results in any [ABI
hashes](../binarycaching.md#abi-hash) changing.
```
